python-checks/src/main/java/org/sonar/python/checks/TooManyReturnsCheck.java

```diff
@@ -42,25 +41,47 @@
   public int max = DEFAULT_MAX;
 
   @Override
-  public Set<AstNodeType> subscribedKinds() {
-    return Collections.singleton(PythonGrammar.FUNCDEF);
-  }
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, ctx -> {
+      PyFunctionDefTree func = ((PyFunctionDefTree) ctx.syntaxNode());
+      ReturnCountVisitor returnCountVisitor = new ReturnCountVisitor();
+      func.body().accept(returnCountVisitor);
 
-  @Override
-  public void visitNode(AstNode node) {
-    AstSelect returnStatements = node.select().descendants(PythonGrammar.RETURN_STMT, PythonGrammar.YIELD_STMT);
-    List<AstNode> returnNodes = new ArrayList<>();
-    for (AstNode returnStatement : returnStatements){
-      if (CheckUtils.insideFunction(returnStatement, node)){
-        returnNodes.add(returnStatement);
+      if (returnCountVisitor.count > max) {
+        PreciseIssue preciseIssue = ctx.addIssue(func.name(), String.format(MESSAGE, returnCountVisitor.count, max));
+
+        returnCountVisitor.yieldStatements.forEach(y -> {
+          preciseIssue.secondary(y, null);
+        });
+        returnCountVisitor.returnStatements.forEach(r -> {
+          preciseIssue.secondary(r, null);
+        });
       }
+    });
+  }
+
+  private static class ReturnCountVisitor extends BaseTreeVisitor {
+
+    private int count = 0;
```

Comment:
Why do we need to store that instead of calling the `size` method of the lists?

Reply:
New commit fixes this.